Add interpolation kwarg to torch.quantile

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49267 Add interpolation kwarg to torch.quantile**

This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

To avoid BC breaking changes, we preserve the original signatures and make the interpolation parameter in the new signatures kwarg only. For now, interpolation cannot have a default value to avoid ambiguity with the deprecated signature. However, due to limitations of codegen and C++, we cannot have a required arg after optional ones. Thus, this PR also makes dim and keepdim requires args. Once we can remove the old signatures, dim, keepdim and interpolation parameters in the new signature will get the default values back.

__TODO__
---
- [x] Run backward compat tests

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D27337117](https://our.internmc.facebook.com/intern/diff/D27337117)

heitorschueroff committed:
Revert "Revert D25428587: [pytorch][PR] add additional interpolation modes for torch.quantile"

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -117,6 +118,12 @@ void quantile_impl(
+  TORCH_CHECK(
+      std::find(interpolations.begin(), interpolations.end(), interpolation) != interpolations.end(),
+      "quantile() interpolation should be one of ",
+      c10::Join(", ", interpolations), ".");
```
" ... but got ..."

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -449,6 +498,10 @@ Tensor quantile(
     double q,
     optional<int64_t> _dim,
     bool keepdim) {
+  TORCH_WARN(
```
TORCH_WARN_ONCE everywhere

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -496,12 +575,156 @@ Tensor nanquantile(
   TORCH_CHECK(
       q >= 0 && q <= 1, "quantile() q must be in the range [0, 1] but got ", q);
   return at::nanquantile(
       self, at::scalar_tensor(q, self.options()), std::move(_dim), keepdim);
```
Is this missing the interpolation param?

heitorschueroff committed:
Update on "Revert "Revert D25428587: [pytorch][PR] add additional interpolation modes for torch.quantile""


Fixed BC breaking change in https://github.com/pytorch/pytorch/pull/48711 and attempting to reland. All implementation credit goes to @RockingJavaBean 

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25541009](https://our.internmc.facebook.com/intern/diff/D25541009)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Revert "Revert D25428587: [pytorch][PR] add additional interpolation modes for torch.quantile""


Fixed BC breaking change in https://github.com/pytorch/pytorch/pull/48711 and attempting to reland. All implementation credit goes to @RockingJavaBean 

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25541009](https://our.internmc.facebook.com/intern/diff/D25541009)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Revert "Revert D25428587: [pytorch][PR] add additional interpolation modes for torch.quantile""


Fixed BC breaking change in https://github.com/pytorch/pytorch/pull/48711 and attempting to reland. All implementation credit goes to @RockingJavaBean 

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25541009](https://our.internmc.facebook.com/intern/diff/D25541009)

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -186,22 +197,38 @@ void quantile_impl(
-  Tensor ranks_above = ranks.ceil_().toType(kLong);
 
+  // adjust ranks based on the interpolation mode
+  if (interpolation == "lower") {
```
FYI, NumPy's interpolation mode implementations can be found here:

https://github.com/numpy/numpy/blob/f4aadabcc0e1968f3b309642af52ef11ae77ca01/numpy/lib/function_base.py#L4004

heitorschueroff replied:
Thank you for the reference. We did follow this but we optimized the cases where we can avoid the interpolation.

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -186,22 +197,38 @@ void quantile_impl(
+    ranks.floor_();
+  } else if (interpolation == "higher") {
+    ranks.ceil_();
+  } else if (interpolation == "nearest") {
```
In the future we should probably add the "midpoint" mode, too

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -414,7 +441,18 @@ Tensor& quantile_out(
+  TORCH_WARN_ONCE(
+      "torch.quantile(input, q, dim=None, keepdim=False, *, out=None) is deprecated "
+      "in favor of the new signature torch.quantile(input, q, dim=None, interpolation='linear', "
+      "keepdim=False, *, out=None) with interpolation parameter.");
```
"with interpolation" -> "with the interpolation"

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -424,13 +462,18 @@ Tensor& quantile_out(
       at::scalar_tensor(q, self.options()),
       std::move(_dim),
+      /*interpolation=*/"linear",
       keepdim);
```
The specification of the ignore_nan parameter is inconsistent across these calls - seems like it should either always be there or always be absent?

heitorschueroff replied:
ignore_nan parameter only exists in the private quantile_impl function. It's not in the public API.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5182,13 +5182,24 @@ def merge_dicts(*dicts):
+By default, :attr:`interpolation` is ``linear`` and :attr:`dim` is ``None`` resulting in the :attr:`input` tensor
 being flattened before computation.
 
+When the quantile value lies between two data points ``i < j``,
```
The `i < j` here seems ambiguous: does it refer to values or indices? An example may be helpful, too. The same ambiguity should be resolved in the first paragraph.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5182,13 +5182,24 @@ def merge_dicts(*dicts):
-Returns the q-th quantiles of each row of the :attr:`input` tensor along the dimension
-:attr:`dim`, doing a linear interpolation when the q-th quantile lies between two
-data points. By default, :attr:`dim` is ``None`` resulting in the :attr:`input` tensor
+Returns the q-th quantiles of each row of the :attr:`input` tensor
```
While we're here, "Returns" -> "Computes"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5182,13 +5182,24 @@ def merge_dicts(*dicts):
-data points. By default, :attr:`dim` is ``None`` resulting in the :attr:`input` tensor
+Returns the q-th quantiles of each row of the :attr:`input` tensor
+along the dimension :attr:`dim` based on :attr:`interpolation`.
+When the desired quantile lies between two data points ``i < j``,
```
This sentence is mostly redundant with the previous one. I think it can be removed.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5182,13 +5182,24 @@ def merge_dicts(*dicts):
+along the dimension :attr:`dim` based on :attr:`interpolation`.
+When the desired quantile lies between two data points ``i < j``,
+the result is computed based on the :attr:`interpolation` value as described below.
+By default, :attr:`interpolation` is ``linear`` and :attr:`dim` is ``None`` resulting in the :attr:`input` tensor
```
This sentence seems like it can be removed, too.

Specifying the default behavior is interesting, but the conflation of the interpolation mode with the flattening behavior of dim here may be confusing (since they're unrelated).

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5182,13 +5182,24 @@ def merge_dicts(*dicts):
 being flattened before computation.
 
+When the quantile value lies between two data points ``i < j``,
+the result is computed according to the given :attr:`interpolation` method as follows:
```
I'm a little confused by how the language works here. Unfortunately mathematical language is often unclear. I think we may even need to revise the first sentence of this doc.

In particular, what is the quantile? There are two numbers here, the specification of the cut, which is something like `.25`, and the value we associate with that cut which has the property that at most 25% of the values in reduced row are below it and at least 75% (rounded up) of the values in the reduced row are above it. Note that it's important we not claim that 25% of values are below it and 75% of values are above. For example, if we have 3 values `{0, 1, 2}` and we take the cut `.8` then the value we associate with that cut will be between `1` and `2` and has the property that 66% of values are below it and 33% are above it. I think this is the right way to think about the computation?

So, back to "what is the quantile?" Is the quantile the specification of the cut, the `.25` or the associated value, like `1.5`, which has the given property? Or is the quantile the pair of them? Or is quantile inherently confusing and the language we use here should just be more specific about about converting a mapping from `[0, 1]` to a set of values with this property?

heitorschueroff replied:
This is a good question. Quantile as Wikipedia defines it are cut points that divide the range of a distribution into segments of equal probability (median is a 2-quantile dividing into 2 segments of 50% probability). However, what this operator computes is a single quantile value such that no more than q elements are smaller than the computed quantile and no more than (1 - q) are bigger than the computed quantile.

heitorschueroff replied:
To answer your question more directly, the quantile is the value that cuts the data. So the 0.5 quantile of an input [1, 2, 3] is 2.

mruberry replied:
OK, so we're finding the quantiles (values) that satisfy the property stipulated by the input. That makes sense.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5182,13 +5182,24 @@ def merge_dicts(*dicts):
+When the quantile value lies between two data points ``i < j``,
+the result is computed according to the given :attr:`interpolation` method as follows:
+
+- ``linear``: ``i + (j - i) * fraction``, where ``fraction`` is the fractional part of the index surrounded by ``i`` and ``j``.
```
"the fractional part of the index surrounded by" is too vague. `fraction` could be replaced with a function of `q` like `q % 1`. Maybe there's a more common way to express that only the decimal or "fractional" part of a number (in base 10) is desired?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5182,13 +5182,24 @@ def merge_dicts(*dicts):
+- ``linear``: ``i + (j - i) * fraction``, where ``fraction`` is the fractional part of the index surrounded by ``i`` and ``j``.
+- ``lower``: ``i``.
+- ``higher``: ``j``.
+- ``nearest``: ``i`` or ``j``, whichever is nearest.
```
"Nearest" is vague because it does not describe what happens with values like 1.5, which is equally "near" 1 and 2 according to the number line we intuitively think of.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5182,13 +5182,24 @@ def merge_dicts(*dicts):
+- ``nearest``: ``i`` or ``j``, whichever is nearest.
+- ``midpoint``: ``(i + j) / 2``.
+
 If :attr:`keepdim` is ``True``, the output dimensions are of the same size as :attr:`input`
```
This paragraph can probably be removed.

In the future we should think about how all reductions should explain the behavior of `keepdim`, and not require individual reductions, like quantile, to bother.

mruberry replied:
The phrase "represents the quantiles" below is vague. We should be more precise about the behavior when `q` is a scalar vs. a 1D tensor.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5222,11 +5235,24 @@ def merge_dicts(*dicts):
+    tensor(2.)
+    >>> torch.quantile(a, 0.6, interpolation='midpoint')
+    tensor(1.5000)
+    >>> torch.quantile(a, 0.6, interpolation='nearest')
```
Probably want to take two quantiles to disambiguate higher and nearest (one nearest goes down, the other goes up).

mruberry commented on torch/_torch_docs.py:
```diff
@@ -5222,11 +5235,24 @@ def merge_dicts(*dicts):
-nanquantile(input, q, dim=None, keepdim=False, *, out=None) -> Tensor
+nanquantile(input, q, dim=None, interpolation='linear', keepdim=False, *, out=None) -> Tensor
 
 This is a variant of :func:`torch.quantile` that "ignores" ``NaN`` values,
```
This doc will need to be updated for consistency with the proposed changes above.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1272,13 +1272,13 @@ def method_tests():
         ('quantile', (S, S, S), (0.5, 0), 'dim', (), [1]),
-        ('quantile', (S, S, S), (0.5, None, True), 'keepdim'),
-        ('quantile', (S, S, S), (0.5, 0, True), 'keepdim_dim', (), [1]),
+        ('quantile', (S, S, S), (0.5, None, 'linear', True), 'keepdim'),
```
In the future we should produce a ReductionInfo class from the OpInfo base class and move quantile and nanquantile to it.

heitorschueroff replied:
Created issue here https://github.com/pytorch/pytorch/issues/49746

mruberry reviewed (commented):
The non-doc changes look pretty good, but I think this is a good opportunity to improve our documentation. We should also run a few more internal tests on this PR. 

Also, add the deprecation label and write a deprecation note at the beginning of this PR explaining what's being deprecated. That will help the release note writer.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -7098,31 +7098,57 @@
   variants: method, function
 
-- func: quantile.scalar_out(Tensor self, float q, int? dim=None, bool keepdim=False, *, Tensor(a!) out) -> Tensor(a!)
+# The following (nan)quantile signatures were deprecated in favor of the ones with interpolation parameter
```
say why.

gchanan replied:
there is a lot of subtlety going on here -- you are remapping the overload name to the "new' function and introducing these deprecated functions.  But are you sure this works for all the use cases?

E.g.:
- Calling with kwargs: if I do something like `quantile(self, q, dim=None, keepdim=False)` isn't that valid for both the deprecated and non-deprecated signatures?  At some point in eager we picked the earliest signature in the list, which would be the wrong thing in this case?  I also don't know what the JIT arg parsing would favor in this case.
- Serializing with TorchScript: do we serialize the overload name and the kwarg properties?
- Same questions with mobile serialization.

gchanan replied:
`quantile(self, q, dim=None, keepdim=False)` does indeed call the deprecated signature.

heitorschueroff replied:
My thinking was that after some time of warning users of the deprecation we can remove the old signature without affecting users of the new signature. The only other option I know is to deprecate it now in which case it will be BC-breaking. Is there another option?

heitorschueroff replied:
In an offline discussion with @mruberry we decided to make the interpolation parameter kwarg only to prevent the BC breaking change.

heitorschueroff committed:
Update on "Revert "Revert D25428587: [pytorch][PR] add additional interpolation modes for torch.quantile""


Fixed BC breaking change in https://github.com/pytorch/pytorch/pull/48711 and attempting to reland. All implementation credit goes to @RockingJavaBean 

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25541009](https://our.internmc.facebook.com/intern/diff/D25541009)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -118,6 +119,16 @@ void quantile_impl(
   TORCH_CHECK(
       self.device() == out.device(),
       "quantile() out tensor must be on the same device as the input tensor");
+  const std::vector<string> interpolations{
```
I'm not sure we want to pass strings to operator kernels (in cases like loss reduction functions to pick an enum in python).  Not a big deal for this PR.

I would recommend we canonicalize the string into an enum as in https://github.com/pytorch/pytorch/blob/1833009202ee5b8e4e25affc325dec87cad87d36/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L768-L779 -- it will be easier to change in the future and you don't have to parse it multiple times.

heitorschueroff committed:
Update on "Revert "Revert D25428587: [pytorch][PR] add additional interpolation modes for torch.quantile""


Fixed BC breaking change in https://github.com/pytorch/pytorch/pull/48711 and attempting to reland. All implementation credit goes to @RockingJavaBean 

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25541009](https://our.internmc.facebook.com/intern/diff/D25541009)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Add interpolation parameter to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

This PR makes the interpolation parameter kwarg only to prevent BC-breaking changes. This is slightly different from NumPy's API however the alternatives were to either break BC or keep conflicting function signatures in native_functions.yaml.

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25877312](https://our.internmc.facebook.com/intern/diff/D25877312)

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -92,14 +92,35 @@ void quick_select_template(
+  } else {
+    TORCH_CHECK(
+        false,
+        "quantile() interpolation must be one of linear, lower, higher, midpoint and nearest, but got ",
```
"and" -> "or"

mruberry commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -412,97 +452,135 @@ Tensor median_impl(const Tensor& self, bool ignore_nan) {
       at::scalar_tensor(q, self.options()),
-      std::move(_dim),
-      keepdim);
+      _dim,
```
This lost a std::move, which isn't a big deal -- any reason why?

heitorschueroff replied:
optional<int64_t> is a trivially copyable value so std::move has no effect.

mruberry reviewed (approved):
Cool. Out of an abundance of caution let's run this through the BC-compat test. It'll be good practice, anyway.

Before merging this, would you make the very small docs tweak and rebase this so we can see the Github CI lint run, too?

heitorschueroff committed:
Update on "Add interpolation parameter to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

This PR makes the interpolation parameter kwarg only to prevent BC-breaking changes. This is slightly different from NumPy's API however the alternatives were to either break BC or keep conflicting function signatures in native_functions.yaml.

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25877312](https://our.internmc.facebook.com/intern/diff/D25877312)

[ghstack-poisoned]

heitorschueroff commented:
> Cool. Out of an abundance of caution let's run this through the BC-compat test. It'll be good practice, anyway.

How can I ran the BC-compat test?

heitorschueroff committed:
Update on "Add interpolation parameter to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

This PR makes the interpolation parameter kwarg only to prevent BC-breaking changes. This is slightly different from NumPy's API however the alternatives were to either break BC or keep conflicting function signatures in native_functions.yaml.

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25877312](https://our.internmc.facebook.com/intern/diff/D25877312)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Add interpolation parameter to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

This PR makes the interpolation parameter kwarg only to prevent BC-breaking changes. This is slightly different from NumPy's API however the alternatives were to either break BC or keep conflicting function signatures in native_functions.yaml.

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25877312](https://our.internmc.facebook.com/intern/diff/D25877312)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Add interpolation kwarg to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

This PR makes the interpolation parameter kwarg only to prevent BC-breaking changes. This is slightly different from NumPy's API however the alternatives were to either break BC or keep conflicting function signatures in native_functions.yaml.

__TODO__
---
- [ ] Run backward compat tests

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25877312](https://our.internmc.facebook.com/intern/diff/D25877312)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Add interpolation kwarg to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

This PR makes the interpolation parameter kwarg only to prevent BC-breaking changes. This is slightly different from NumPy's API however the alternatives were to either break BC or keep conflicting function signatures in native_functions.yaml.

__TODO__
---
- [ ] Run backward compat tests

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25877312](https://our.internmc.facebook.com/intern/diff/D25877312)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Add interpolation kwarg to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

To avoid BC breaking changes, we preserve the original signature and make the interpolation parameter kwarg only. For now, interpolation cannot have a default value to avoid ambiguity with the deprecated signature. However, due to limitations of codegen and C++, we cannot have a required arg after optional ones. Thus, this PR also moves dim and keepdim as kwarg only after the interpolation parameter.

__TODO__
---
- [ ] Run backward compat tests

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25877312](https://our.internmc.facebook.com/intern/diff/D25877312)

[ghstack-poisoned]

matthewfeickert commented:
@heitorschueroff thanks very much for working on this PR &mdash; it is really great to see, and we appreciate the work that you've done here and its impact on https://github.com/scikit-hep/pyhf/issues/815. :bow: 

@heitorschueroff @mruberry To follow up, out of curiosity and the enjoyment of seeing this in a future release, on the last point here

> How can I ran the BC-compat test?

has there been a resolution on this?

heitorschueroff commented:
> @heitorschueroff thanks very much for working on this PR — it is really great to see, and we appreciate the work that you've done here and its impact on [scikit-hep/pyhf#815](https://github.com/scikit-hep/pyhf/issues/815). 🙇
> 
> @heitorschueroff @mruberry To follow up, out of curiosity and the enjoyment of seeing this in a future release, on the last point here
> 
> > How can I ran the BC-compat test?
> 
> has there been a resolution on this?

@matthewfeickert Yes, we have found a resolution for it which will be working on in the coming week. Thank you for letting us know of the community needs. As a side note, we have discussed adding percentile to PyTorch before but ultimately decided that quantile offers the same capabilities.

matthewfeickert commented:
> Yes, we have found a resolution for it which will be working on in the coming week.

:rocket: Fantastic. Thank you very much for the update as well!

> As a side note, we have discussed adding percentile to PyTorch before but ultimately decided that quantile offers the same capabilities.

Yup! I remember reading those discussions and all the arguments made total sense to me. We (`pyhf`) just wrap `quantile` to still provide `percentile`.

heitorschueroff committed:
Update on "Add interpolation kwarg to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

To avoid BC breaking changes, we preserve the original signature and make the interpolation parameter kwarg only. For now, interpolation cannot have a default value to avoid ambiguity with the deprecated signature. However, due to limitations of codegen and C++, we cannot have a required arg after optional ones. Thus, this PR also moves dim and keepdim as kwarg only after the interpolation parameter.

__TODO__
---
- [ ] Run backward compat tests

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25877312](https://our.internmc.facebook.com/intern/diff/D25877312)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Add interpolation kwarg to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

To avoid BC breaking changes, we preserve the original signatures and make the interpolation parameter in the new signatures kwarg only. For now, interpolation cannot have a default value to avoid ambiguity with the deprecated signature. However, due to limitations of codegen and C++, we cannot have a required arg after optional ones. Thus, this PR also makes dim and keepdim requires args. Once we can remove the old signatures, dim, keepdim and interpolation parameters in the new signature will get the default values back.

__TODO__
---
- [ ] Run backward compat tests

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25877312](https://our.internmc.facebook.com/intern/diff/D25877312)

[ghstack-poisoned]

heitorschueroff commented:
Update on the state of this PR: The scripts used for running the backward compat check seem to be broken, I'm working with the owner of the script to resolve it.

heitorschueroff committed:
Update on "Add interpolation kwarg to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

To avoid BC breaking changes, we preserve the original signatures and make the interpolation parameter in the new signatures kwarg only. For now, interpolation cannot have a default value to avoid ambiguity with the deprecated signature. However, due to limitations of codegen and C++, we cannot have a required arg after optional ones. Thus, this PR also makes dim and keepdim requires args. Once we can remove the old signatures, dim, keepdim and interpolation parameters in the new signature will get the default values back.

__TODO__
---
- [ ] Run backward compat tests

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D25877312](https://our.internmc.facebook.com/intern/diff/D25877312)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Add interpolation kwarg to torch.quantile"


This PR builds upon the PR https://github.com/pytorch/pytorch/pull/48711 by @RockingJavaBean. The original PR introduced a BC breaking change by making the interpolation parameter positional. Thus, previous invocations of torch.quantile that did not include the interpolation parameter failed after the PR landed. 

To avoid BC breaking changes, we preserve the original signatures and make the interpolation parameter in the new signatures kwarg only. For now, interpolation cannot have a default value to avoid ambiguity with the deprecated signature. However, due to limitations of codegen and C++, we cannot have a required arg after optional ones. Thus, this PR also makes dim and keepdim requires args. Once we can remove the old signatures, dim, keepdim and interpolation parameters in the new signature will get the default values back.

__TODO__
---
- [x] Run backward compat tests

This reverts commit 2f1d1eb7df5e8032392b73751c84025a2aa3d1ee.

Differential Revision: [D27337117](https://our.internmc.facebook.com/intern/diff/D27337117)

[ghstack-poisoned]

mruberry reviewed (approved):
Still looks good

matthewfeickert commented:
Woo! Thanks so much for your hard work here @heitorschueroff &mdash; super appreciated! Looking forward to seeing this in an upcoming release. :+1:
